<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from fourth_robot.urdf.xacro        | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="rs_robot" xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- =============== Link & Joint =============== -->
  <!-- Base -->
  <link name="base_footprint"/>
  <joint name="base_link_joint" type="fixed">
    <origin xyz="0 0 0"/>
    <parent link="base_footprint"/>
    <child link="base_link"/>
  </joint>
  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/base/base_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/base/base_link.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0.017"/>
      <mass value="1.2"/>
      <inertia ixx="0.0001873" ixy="0" ixz="0" iyy="0.000055" iyz="0" izz="0.0002283"/>
    </inertial>
  </link>
  <joint name="fr_steering_joint" type="revolute">
    <origin xyz="0.077 -0.031 -0.004"/>
    <parent link="base_link"/>
    <child link="fr_steering_link"/>
    <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="fr_steering_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/suspension/front_right_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/suspension/sus_link.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 -0.02"/>
      <mass value="0.15"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.00001" iyz="0" izz="0.000004"/>
    </inertial>
  </link>
  <joint name="fl_steering_joint" type="revolute">
    <origin xyz="0.077 0.031 -0.004"/>
    <parent link="base_link"/>
    <child link="fl_steering_link"/>
    <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="fl_steering_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/suspension/front_left_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/suspension/sus_link.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 -0.02"/>
      <mass value="0.15"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.00001" iyz="0" izz="0.000004"/>
    </inertial>
  </link>
  <joint name="rr_steering_joint" type="revolute">
    <origin xyz="-0.077 -0.031 -0.004"/>
    <parent link="base_link"/>
    <child link="rr_steering_link"/>
    <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="rr_steering_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/suspension/rear_right_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/suspension/sus_link.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 -0.02"/>
      <mass value="0.15"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.00001" iyz="0" izz="0.000004"/>
    </inertial>
  </link>
  <joint name="rl_steering_joint" type="revolute">
    <origin xyz="-0.077 0.031 -0.004"/>
    <parent link="base_link"/>
    <child link="rl_steering_link"/>
    <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="rl_steering_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/suspension/rear_left_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/suspension/sus_link.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 -0.02"/>
      <mass value="0.15"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.00001" iyz="0" izz="0.000004"/>
    </inertial>
  </link>
  <joint name="fr_wheel_joint" type="continuous">
    <origin xyz="0 -0.021 -0.025"/>
    <parent link="fr_steering_link"/>
    <child link="fr_wheel_link"/>
    <limit effort="1.5" velocity="20"/>
    <axis xyz="0 1 0"/>
  </joint>
  <link name="fr_wheel_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/wheel/wheel_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/wheel/robot_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.0 -0.007 -0.0"/>
      <mass value="0.05"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.000018" iyz="0" izz="0.00001"/>
    </inertial>
  </link>
  <joint name="fl_wheel_joint" type="continuous">
    <origin rpy="0 0 3.14159274" xyz="0 0.021 -0.025"/>
    <parent link="fl_steering_link"/>
    <child link="fl_wheel_link"/>
    <limit effort="1.5" velocity="20"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <link name="fl_wheel_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/wheel/wheel_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/wheel/robot_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.0 -0.007 -0.0"/>
      <mass value="0.05"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.000018" iyz="0" izz="0.00001"/>
    </inertial>
  </link>
  <joint name="rr_wheel_joint" type="continuous">
    <origin xyz="0 -0.021 -0.025"/>
    <parent link="rr_steering_link"/>
    <child link="rr_wheel_link"/>
    <limit effort="1.5" velocity="20"/>
    <axis xyz="0 1 0"/>
  </joint>
  <link name="rr_wheel_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/wheel/wheel_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/wheel/robot_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.0 -0.007 -0.0"/>
      <mass value="0.05"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.000018" iyz="0" izz="0.00001"/>
    </inertial>
  </link>
  <joint name="rl_wheel_joint" type="continuous">
    <origin rpy="0 0 3.14159274" xyz="0 0.021 -0.025"/>
    <parent link="rl_steering_link"/>
    <child link="rl_wheel_link"/>
    <limit effort="1.5" velocity="20"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <link name="rl_wheel_link">
    <visual>
      <geometry>
        <mesh filename="package://robot_description/meshes/DAE/wheel/wheel_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://robot_description/meshes/STL/wheel/robot_wheel.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.0 -0.007 -0.0"/>
      <mass value="0.05"/>
      <inertia ixx="0.00001" ixy="0" ixz="0" iyy="0.000018" iyz="0" izz="0.00001"/>
    </inertial>
  </link>
  <!-- Box alargado en horizontal -->
  <link name="box_central">
    <origin xyz="0 0 0.35" rpy="0 1.5708 1.5708"/>
    <collision name="collision">
      <geometry>
        <cylinder length="0.4" radius="0.1"/>
      </geometry>
    </collision>
    <!-- <visual name="visual">
      <geometry>
        <box size="0.2 0.125 0.125"/>
      </geometry>
      <material>
        <ambient>0.8 0 0 1</ambient>
        <diffuse>1 0 0 1</diffuse>
        <specular>1 0.2 0.2 1</specular>
        <emissive>0 0 0 0</emissive>
      </material>
    </visual> -->
  </link>
  <!-- Box plano vertical izquierdo -->
  <link name="box_plano_izquierdo">
    <origin xyz="0 -0.2 0.35" rpy="0 0 1.5708"/> 
    <collision name="collision">
      <geometry>
        <box size="0.025 0.25 0.35"/>
      </geometry>
    </collision>
    <!-- <visual name="visual">
      <geometry>
        <box size="0.025 0.25 0.25"/>
      </geometry>
      <material>
        <ambient>0.8 0 0 1</ambient>
        <diffuse>1 0 0 1</diffuse>
        <specular>1 0.2 0.2 1</specular>
        <emissive>0 0 0 0</emissive>
      </material>
    </visual> -->
  </link>
  <!-- Box plano vertical derecho -->
  <link name="box_plano_derecho">
    <origin xyz="0 0.2 0.35" rpy="0 0 1.5708"/>
    <collision name="collision">
      <geometry>
        <box size="0.025 0.25 0.35"/>
      </geometry>
    </collision>
    <!-- <visual name="visual">
      <geometry>
        <box size="0.025 0.25 0.25"/>
      </geometry>
      <material>
        <ambient>0.8 0 0 1</ambient>
        <diffuse>1 0 0 1</diffuse>
        <specular>1 0.2 0.2 1</specular>
        <emissive>0 0 0 0</emissive>
      </material>
    </visual> -->
  </link>
  <!-- Uni贸n revoluta izquierda -->
  <joint name="joint_izquierdo" type="revolute">
    <parent link="box_plano_izquierdo"/>
    <child link="box_central"/>
    <origin xyz="0 0 0" rpy="0 0 1.5708"/>
    <axis xyz="0 1 0"/>
    <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
  </joint>
  <!-- Uni贸n revoluta derecha -->
  <joint name="joint_derecho" type="revolute">
    <parent link="box_plano_derecho"/>
    <child link="box_central"/>
    <origin xyz="0 0 0" rpy="0 0 1.5708"/>
    <axis xyz="0 1 0"/>
    <limit effort="5" lower="-2.1" upper="2.1" velocity="6.28"/>
  </joint>
  <!-- Uni贸n fixed derecha al UGV -->
  <joint name="joint_derecho_to_ugv" type="fixed">
    <parent link="base_link"/>
    <child link="box_plano_derecho"/>
    <origin xyz="0 0 0" rpy="0 0 1.5708"/> 
  </joint>
  <!-- Uni贸n fixed izquierda al UGV -->
  <joint name="joint_izquierdo_to_ugv" type="fixed">
    <parent link="base_link"/>
    <child link="box_plano_izquierdo"/>
    <origin xyz="0 0 0" rpy="0 0 1.5708"/> 
  </joint>
  
  
  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    
    <joint name="fl_wheel_joint">
      <command_interface name="velocity">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="fr_wheel_joint">
      <command_interface name="velocity">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="rl_wheel_joint">
      <command_interface name="velocity">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="rr_wheel_joint">
      <command_interface name="velocity">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="joint_izquierdo">
      <command_interface name="velocity">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="fl_steering_joint">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="fr_steering_joint">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="rl_steering_joint">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="rr_steering_joint">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <!-- <joint name="joint_izquierdo">
      <command_interface name="position">
        <param name="min">-3.14</param>
        <param name="max">3.14</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint> -->

  </ros2_control>

</robot>

